cmd/contour/serve.go

```diff
@@ -372,50 +367,80 @@ func doServe(log logrus.FieldLogger, ctx *serveContext) error {
 		Logger:    log.WithField("context", "dynamicHandler"),
 	}
 
-	// Register our resource event handler with the k8s informers,
-	// using the SyncList to keep track of what to sync later.
-	var informerSyncList k8s.InformerSyncList
+	// Inform on DefaultResources.
+	for _, r := range k8s.DefaultResources() {
+		inf, err := clients.InformerForResource(r)
+		if err != nil {
+			log.WithError(err).WithField("resource", r).Fatal("failed to create informer")
+		}
 
-	// Inform on DefaultResources
-	informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.DefaultResources()...)
+		inf.AddEventHandler(dynamicHandler)
+	}
 
+	// Inform on service-apis types if they are present.
 	if ctx.UseExperimentalServiceAPITypes {
-		// Check if the resource exists in the API server before setting up the informer.
-		if !clients.ResourcesExist(k8s.ServiceAPIResources()...) {
-			log.WithField("InformOnResources", "ExperimentalServiceAPITypes").Warnf("resources %v not found in api server", k8s.ServiceAPIResources())
-		} else {
-			informerSyncList.InformOnResources(clusterInformerFactory, dynamicHandler, k8s.ServiceAPIResources()...)
+		for _, r := range k8s.ServiceAPIResources() {
```

Comment:
Might help for readability to move this into a method to clean up the serve method.